proof/refine/RISCV64/IpcCancel_R.thy

```diff
@@ -1051,7 +1051,8 @@ lemma scReplies_of_scTCB_update[simp]:
 crunch schedContextDonate
   for replies_of': "\<lambda>s. P (replies_of' s)" (* this interfers with list_refs_of_replies' *)
   and scReplies_of[wp]: "\<lambda>s. P' (scReplies_of s)"
-  (simp: crunch_simps wp: crunch_wps)
+  (simp: crunch_simps wp: crunch_wps
+   rule: schedContextDonate_def[simplified updateSchedContext_def])
```

Comment:
This use of `rule` feels a little bit strange, is it to avoid proving the equivalent lemmas for `updateSchedContext`?

Reply:
Some of the changes in Refine are a bit of a hack job, yes. A bit of background...

I wanted to use `updateSchedContext` rather than `setSchedContext` because firstly, `updateSchedContext` does seem to be a better abstraction to use (where we get the sched context object and then immediately set it with a variant of the object we get) and second, we already have `ccorres` rules for `updateSchedContext` which are exact analogues of the preexisting rules for `threadSet`, and I wanted to just use those.

But `updateSchedContext` was only defined fairly late on in the Refine run, and it has quite different rules when compared with the rules for `setSchedContext`. For example, some of the `updateSchedContext` rules will use the `opt_pred/opt_map` style functions for heap access, rather than `obj_at'` style, and this causes the proofs to go haywire unless we just try to get it back into the original formulation. So sometimes I do properly update the proofs for this change, and other times I just try to get back what we had before. Maybe once the `opt_pred` style is the only style we have, we could make these clunky proofs nicer

Reply:
This could use at least a short comment then ("prefer X because Y", or "avoid X due to Y") or something like that. This kind of hidden knowledge is what makes proofs difficult to figure out / fix / update later, e.g. in the case you cited if we ever want to go forth and change to the `opt_pred` style.

Reply:
Right, that all makes sense, I actually just added an `updateReply` for that exact reason of it being nicer to work with.

I can definitely understand making the minimal change to keep the proofs working. A comment could be helpful if you can find a suitable spot for it.

Reply:
On second thoughts, I decided to go with writing the rule

```
lemma updateSchedContext_scReplies_of:
  "(⋀sc. scReply (f sc) = scReply sc) ⟹ updateSchedContext scPtr f ⦃λs. P' (scReplies_of s)⦄"
```

and then adding that to the `crunch` instead. Hopefully that's a nicer approach

Reply:
Yeah, that's a lot nicer, I'm happy with that. Do you definitely need to add it to the `crunch` though? I would have guessed that it picked it up automatically.

Reply:
You're right, it's not needed in the `crunch`. After all these years, I thought it had to be in the `wp` for `crunch` to pick it up. I'll remove it

Reply:
`crunch` picks rules up in a variety of different ways, including if the rule is already in the `wp` set and solves the generated goal, or if it has the same name as what `crunch` would produce. The former is how it works  when a rule is passed in via `wp`, the latter is why you don't need to do anything in this case.

Reply:
That's great to know, thanks!